### ISSUE TITLE
[REVIEW]: PyNomaly

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#845]
**Submitting author:** @vc1492a (<a href="http://orcid.org/0000-0002-5279-4143">Valentino Constantinou</a>)
**Repository:** <a href="https://www.github.com/vc1492a/PyNomaly" target ="_blank">https://www.github.com/vc1492a/PyNomaly</a>
**Version:** v0.2.4
**Editor:** @Kevin-Mattheus-Moerman
**Reviewers:** @llllllllll
**Archive:** <a href="https://doi.org/10.5281/zenodo.1472519" target="_blank">10.5281/zenodo.1472519</a>

## Status

[![status](http://joss.theoj.org/papers/f4d2cfe680768526da7c1f6a2c103266/status.svg)](http://joss.theoj.org/papers/f4d2cfe680768526da7c1f6a2c103266)

Status badge code:

```
HTML: <a href="http://joss.theoj.org/papers/f4d2cfe680768526da7c1f6a2c103266"><img src="http://joss.theoj.org/papers/f4d2cfe680768526da7c1f6a2c103266/status.svg"></a>
Markdown: [![status](http://joss.theoj.org/papers/f4d2cfe680768526da7c1f6a2c103266/status.svg)](http://joss.theoj.org/papers/f4d2cfe680768526da7c1f6a2c103266)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://www.github.com/vc1492a/PyNomaly" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@yzhao062, @llllllllll, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.theoj.org/about#reviewer_guidelines. Any questions/concerns please let @Kevin-Mattheus-Moerman know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://www.github.com/vc1492a/PyNomaly">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** Does the release version given match the GitHub release (v0.2.0)?
- [x] **Authorship:** Has the submitting author (@vc1492a) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

---

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://www.github.com/vc1492a/PyNomaly">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** Does the release version given match the GitHub release (v0.2.0)?
- [x] **Authorship:** Has the submitting author (@vc1492a) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

**Comment by whedon** [comment]
Hello human, I'm @whedon. I'm here to help you with some common editorial tasks. @yzhao062 it looks like you're currently assigned as the reviewer for this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[--> Check article proof :page_facing_up: <--](https://github.com/openjournals/joss-papers/blob/joss.00845/joss.00845/10.21105.joss.00845.pdf)

**Comment by Kevin-Mattheus-Moerman** [comment]
@vc1492a @yzhao062 this is where the review takes place. I aim to find more reviewers but to avoid further delay we can get started here. 
@yzhao062 could you provide an estimate as to when you think you'll be able to finish the review? Thanks! :robot: :rocket:

**Comment by vc1492a** [comment]
@yzhao062 Current GitHub release is 0.2.1, as PyNomaly received an update after submission.

**Comment by yzhao062** [comment]
@Kevin-Mattheus-Moerman @vc1492a I am about to start by following the review schedule  above. Given I need some time to familiarize myself with the process, I would say 7-10 days is a safe estimate (should be faster though).

**Comment by Kevin-Mattheus-Moerman** [comment]
@yzhao062 thanks. Okay, let me know if you need help.

**Comment by yzhao062** [comment]
Hi @Kevin-Mattheus-Moerman @vc1492a ,

I have followed the tasklist to conduct the first round review and marked most of the items as done. There are 2 items unchecked and want to list them below:

1. **Functionality documentation**: the parameters are introduced in readme.MD. Given there are only 2 parameters to tweak (extent and n_neighbors), it should be fine without a full documentation (e.g. readthedocs). However, it may be helpful to open another section to list the the default values (0.997 and 10) and your recommendations for picking the parameters.

2. **Automated tests**:  it could be helpful to include some basic tests to make sure the model is running. For example, it is easy to generate some pseudo data and use loOP to do a classification and verify the result, which could at least make sure it is doing the job on the simplest data. Feel free to use the [data_generation](https://github.com/yzhao062/Pyod/blob/master/pyod/utils/data.py) model here if that saves some time.

Other than those two items, I think it is a very solid and useful package to go which I used in my last paper. Since I am a first-time reviewer, please advise what is the best way to address the above two tasks.

Thanks,
Yue

**Comment by Kevin-Mattheus-Moerman** [comment]
@vc1492a how are you getting on? Are you able to respond to the issues raised by @yzhao062 ? Thanks

**Comment by vc1492a** [comment]
@Kevin-Mattheus-Moerman Thanks for the message. I was under the impression that the work would be reviewed by a second reviewer prior to moving to the revision process. As such, I have not yet started on the recommendations @yzhao062 provided above but please let me know whether I should move forward with the revision process. I can get started in the next few days, and should be able to provide solutions to the above recommendations in about two weeks time.

**Comment by Kevin-Mattheus-Moerman** [comment]
@vc1492a I am searching for a second reviewer but please do start working on the revision process already.

**Comment by vc1492a** [comment]
@Kevin-Mattheus-Moerman Thanks for reply. Will get started on the revision process and will get back in touch once I have pushed the suggested changes.

**Comment by vc1492a** [comment]
@Kevin-Mattheus-Moerman I have addressed the issues raised by @yzhao062. Additional documentation for parameter has been added to the readme, and a series of automated tests are available in the `tests` directory (using `pytest`). Additionally, examples have been added to the `examples` directory. Please let me know if there's any additional items I should address.

**Comment by Kevin-Mattheus-Moerman** [comment]
@llllllllll is this something you'd be interested in reviewing?

**Comment by llllllllll** [comment]
I am interested in this, but I am moving tomorrow and will need a few days to unpack. I can look at this in a week or so.

**Comment by Kevin-Mattheus-Moerman** [comment]
@llllllllll great thanks. I'll add you as reviewer and look forward to seeing your comments. All the best with the move :house_with_garden:

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon assign @llllllllll as reviewer

**Comment by whedon** [comment]
OK, the reviewer is @llllllllll

**Comment by Kevin-Mattheus-Moerman** [comment]
@yzhao062 could you respond to how @vc1492a has dealt with your feedback? Thanks

**Comment by yzhao062** [comment]
@Kevin-Mattheus-Moerman will do a review in 3 days for sure:)

**Comment by yzhao062** [comment]
@Kevin-Mattheus-Moerman I have successfully run the newly added tests. I am confident it is a sound and useful package! Great work @vc1492a ! I think it is good to go:)

**Comment by Kevin-Mattheus-Moerman** [comment]
Thanks @yzhao062 for your end of the review work! :tada:

**Comment by Kevin-Mattheus-Moerman** [comment]
@vc1492a we will now await @llllllllll review comments.

**Comment by Kevin-Mattheus-Moerman** [comment]
@llllllllll could you give us an estimate as to when you expect to work on this review process? Thanks

**Comment by Kevin-Mattheus-Moerman** [comment]
@llllllllll could you give us an estimate as to when you expect to work on this review process? :rabbit: :rose: :robot:

**Comment by llllllllll** [comment]
Sorry, really let this slip. I will review this tonight.

**Comment by llllllllll** [comment]
Sorry it took me so long to get to this. This is a really cool project.

It looks like you just merged my install pr, so I will check that off. Thanks! I had one small comment about clarifying the layout of the input. It only took me a minute to figure it out once I played with it, but it made it hard to follow the examples.

For testing I used daily returns of US equities for the year of 2016. I played around with different ways of extracting features. I also tried using sector labels as my cluster labels. I was very happy to see that this library identified some known data quality issues, mostly around missing splits. More importantly, it found a few new issues. I will definitely try to use this more in the future. Thanks again, and sorry for taking so long on this.

**Comment by vc1492a** [comment]
@llllllllll Thanks for your comments and suggestions, and thanks for submitting the pr. I'll take a look at your opened issue regarding the layout of the input and will see what changes I can make to PyNomaly and/or the documentation to improve clarification.

**Comment by vc1492a** [comment]
@Kevin-Mattheus-Moerman I have added additional clarification in `readme.md` following @llllllllll's review and comments. Please let me know if there are any additional steps I should take, thanks!

**Comment by Kevin-Mattheus-Moerman** [comment]
Thanks @vc1492a. @llllllllll can you please check out these new developments.

**Comment by llllllllll** [comment]
That looks great, thank you!

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.00845/joss.00845/10.21105.joss.00845.pdf)

**Comment by Kevin-Mattheus-Moerman** [comment]
This is looking good now. Thanks @yzhao062, @llllllllll for your review efforts! :rocket: 

@vc1492a at this point could you please: 
1) Inform us if the version number (v0.2.0) assigned here is still accurate or if you've since created a new release? 
2) Deposit an archived version of the reviewed software in a service like [Zenodo](https://zenodo.org/) (see also: https://guides.github.com/activities/citable-code/) and report back with the DOI of the archived version.

**Comment by vc1492a** [comment]
@Kevin-Mattheus-Moerman Thanks for the update. The current version number is v0.2.4 - there's been a few releases since the start of the review process. I've deposited an archived version of the software, it's DOI is `10.5281/zenodo.1472519`. Besides the version number, the article proof looks good to me as well. Thanks @yzhao062 and @llllllllll for the review!

**Comment by Kevin-Mattheus-Moerman** [comment]
@arfon I recommend we accept this work in JOSS. Above is the archive DOI. Please also note the version number change. Thanks.

**Comment by arfon** [comment]
@whedon set 10.5281/zenodo.1472519 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.1472519" target="_blank">10.5281/zenodo.1472519</a> is the archive.

**Comment by arfon** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/38

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/38, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by arfon** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/39
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.00845](https://doi.org/10.21105/joss.00845)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? notify your editorial technical team...

**Comment by arfon** [comment]
@llllllllll - many thanks for your review here and to @Kevin-Mattheus-Moerman for editing this submission ✨ 

@vc1492a - your paper is now accepted into JOSS and your DOI is https://doi.org/10.21105/joss.00845 :zap: :rocket: :boom:

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](http://joss.theoj.org/papers/10.21105/joss.00845/status.svg)](https://doi.org/10.21105/joss.00845)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.00845">
  <img src="http://joss.theoj.org/papers/10.21105/joss.00845/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: http://joss.theoj.org/papers/10.21105/joss.00845/status.svg
   :target: https://doi.org/10.21105/joss.00845
```
This is how it will look in your documentation:

[![DOI](http://joss.theoj.org/papers/10.21105/joss.00845/status.svg)](https://doi.org/10.21105/joss.00845)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: http://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://www.flipcause.com/secure/cause_pdetails/Mjk3Nzk=
